securesystemslib/spx_keys.py

```diff
@@ -0,0 +1,228 @@
+# public/private keys are hexlified.
+import binascii
+
+# 'os' required to generate OS-specific randomness (os.urandom) suitable for
+# cryptographic use.
+# http://docs.python.org/2/library/os.html#miscellaneous-functions
+import os
+
+# Import the pyspx library, if available. This library is required to use
+# spx signatures.
+import pyspx.shake256_192s as pyspx
+
+import securesystemslib.formats
+import securesystemslib.exceptions
+import securesystemslib.schema as SCHEMA
+
+# Define lengths of SPX keys and signature bytes
+# NOTE: Define module scope schemas here to avoid conditional imports of
+# optional 'pyspx' package in 'formats' module. ImportError and IOError should
+# be handled by whoever imports this 'spx_keys' module.
+SPX_PUBLIC_BYTES_SCHEMA = SCHEMA.LengthBytes(pyspx.crypto_sign_PUBLICKEYBYTES)
+SPX_PRIVATE_BYTES_SCHEMA = SCHEMA.LengthBytes(pyspx.crypto_sign_SECRETKEYBYTES)
+SPX_SIG_BYTES_SCHEMA = SCHEMA.LengthBytes(pyspx.crypto_sign_BYTES)
+
+def generate_public_and_private():
+  """
+  <Purpose>
+    Generate a pair of spx public and private keys with pyspx.  The public
+    and private keys returned conform to 'SPX_PUBLIC_BYTES_SCHEMA' and
+    'SPX_PRIVATE_BYTES_SCHEMA', respectively.
+
+    An spx seed key is a random 128-byte string. Public keys are 64 bytes.
+
+    >>> public, private = generate_public_and_private()
+    >>> SPX_PUBLIC_BYTES_SCHEMA.matches(public)
+    True
+    >>> SPX_PRIVATE_BYTES_SCHEMA.matches(private)
+    True
+
+  <Arguments>
+    None.
+
+  <Exceptions>
+    NotImplementedError, if a randomness source is not found by 'os.urandom'.
+
+  <Side Effects>
+    The spx keys are generated by first creating a random seed
+    with os.urandom() and then calling pyspx's pyspx.signing.SigningKey().
+
+  <Returns>
+    A (public, private) tuple that conform to
+    'SPX_PUBLIC_BYTES_SCHEMA' and
+    'SPX_PRIVATE_BYTES_SCHEMA', respectively.
+  """
+
+  # Generate spx's seed key by calling os.urandom().  The random bytes
+  # returned should be suitable for cryptographic use and is OS-specific.
+  # Raise 'NotImplementedError' if a randomness source is not found.
+  seed = os.urandom(pyspx.crypto_sign_SEEDBYTES)
+
+  # Generate the public key.  pyspx performs the actual key generation.
+  public, private = pyspx.generate_keypair(seed)
+
+  return public, private
+
+
+
+
+
+def create_signature(private_key, data, scheme):
+  """
+  <Purpose>
+    Return a (signature, scheme) tuple, where the signature scheme is 'spx'
+    and is always generated by pyspx.  The signature returned
+    conforms to 'SPX_SIG_BYTES_SCHEMA'.
+
+    >>> public, private = generate_public_and_private()
+    >>> data = b'The quick brown fox jumps over the lazy dog'
+    >>> scheme = 'spx'
+    >>> signature, scheme = \
+        create_signature(private, data, scheme)
+    >>> SPX_SIG_BYTES_SCHEMA.matches(signature)
+    True
+    >>> scheme == 'spx'
+    True
+    >>> signature, scheme = \
```

Comment:
Are these lines repeated?

Reply:
Good eyes. Seems like a copy-pasta error from `ed25519_keys.py`, which apparently served as base for `spx_keys.py`. Fixed in f6e47b9.